### PR TITLE
ENH: Update eigen3.4

### LINKED ISSUE #3045: Update eigen to 3.4
From the [look of the branch name](https://github.com/InsightSoftwareConsortium/ITK/blob/51dc8fee7da4e1a52d0190327316da512aac84f0/Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh#L7), ITK master currently uses eigen 3.3.x.

3.4 was [released](https://eigen.tuxfamily.org/index.php?title=3.4) a few months ago.

In order to eventually fix all the [-Wreserved-id-macro warnings](https://open.cdash.org/viewBuildError.php?type=1&buildid=7660676) eigen will eventually need updating.  So I figure at least moving to the latest release will be a first step.

### PR DESCRIPTION
**Pull request by phcerdan**
Fixes #3045

### CODE REVIEW

**Comment by seanm** [comment]
Looks reasonable (not that I'll pretend to actually have reviewed all the eigen changes!)

**Comment by phcerdan** [comment]
It's ready. I had to do a manual merge (copy files from upstream manually) for the Eigen/src/Geometry module, it was keeping old versions instead of new after conflicts. Not sure why.

I tested against an external module using Eigen (ITKTotalVariation) and works ok.

**Comment by dzenanz** [comment]
The [update script](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh) should be updated. Take a look at a [recent example](https://github.com/InsightSoftwareConsortium/ITK/commit/fc46ec6bcefa1e1154a6524cfcd7eb36857a51cd). The new version works better, produces way less conflicts. Especially it does not produce stupid, avoidable conflicts.

File: Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@ thirdparty_module_name="Eigen3"
 
 # readonly repo="https://gitlab.kitware.com/third-party/eigen.git"
 upstream_git_url="https://gitlab.kitware.com/phcerdan/eigen.git"
-upstream_git_branch="for/itk-3.3-find_package"
+upstream_git_branch="for/itk-3.4"
```
Comment (dzenanz) [inline Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh:7]: Changes which are not part of Eigen's upstream changes should be in separate commits, such as changes to this file. A recent example:
https://github.com/InsightSoftwareConsortium/ITK/pull/2832

**Review by dzenanz — COMMENTED** [review]
Mostly looks good.

**Comment by phcerdan** [comment]
> The [update script](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh) should be updated. Take a look at a [recent example](https://github.com/InsightSoftwareConsortium/ITK/commit/fc46ec6bcefa1e1154a6524cfcd7eb36857a51cd). The new version works better, produces way less conflicts. Especially it does not produce stupid, avoidable conflicts.

I am confused. There is a commit explicitly changing my original UpdateFromStream from what you call the new version to old version: 06d395f4bbb @vfonov 

> Changes which are not part of Eigen's upstream changes should be in separate commits, such as changes to this file

Thanks, will do.

File: Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh [hunk 2]
```diff
@@ -7,12 +7,10 @@ shopt -s dotglob
 readonly name="Eigen3"
 readonly ownership="Eigen Upstream <kwrobot@kitware.com>"
 readonly subtree="Modules/ThirdParty/Eigen3/src/itkeigen"
-# readonly repo="https://github.com/eigenteam/eigen-git-mirror"
```
Comment (dzenanz) [inline Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh]: Leaving the official mirror here will probably save somebody some searching.

File: Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh [hunk 3]
```diff
@@ -7,12 +7,10 @@ shopt -s dotglob
...
 # readonly repo="https://gitlab.kitware.com/third-party/eigen.git"
 readonly repo="https://gitlab.kitware.com/phcerdan/eigen.git"
-readonly tag="for/itk-eigen-3.3"
+readonly tag="for/itk-3.4"
 readonly paths="
-Eigen/CMakeLists.txt
```
Comment (dzenanz) [inline Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh]: Why remove automated updates for this file?

**Review by dzenanz — COMMENTED** [review]
More questions again :(

File: Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh (hunk 2 above)
Comment (phcerdan) [reply Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh]: That's was the non-official mirror when eigen was in mercurial. Now they have a proper git version here:
https://gitlab.com/libeigen/eigen.

I will put it there.

File: Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh (hunk 3 above)
Comment (phcerdan) [reply Modules/ThirdParty/Eigen3/UpdateFromUpstream.sh]: They removed it upstream, the Update script was failing because it.

**Review by dzenanz — APPROVED** [review]

**Comment by seanm** [comment]
Not too surprisingly this added some colour in cdash... I'll take a look at it...

**Comment by phcerdan** [comment]
> Not too surprisingly this added some colour in cdash... I'll take a look at it...

Just Yellow or RED :fearful:? 

Thanks for reporting upstream, we can update when they fix it. This Update was on top of their current 3.4 branch.

**Comment by phcerdan** [comment]
Fixing warning `Wused-but-marked-unused` at #3095
